Word Noise Transformation

saqibns opened the pull request:
This transformation chooses a set of words at random from the context and the question and forms a sentence out of them. The sentence is then prepended or appended to the context.

It is inspired by the **AddAny** method described in [Adversarial SQUAD](https://arxiv.org/abs/1707.07328).

kaustubhdhole commented on transformations/word_noise/transformation.py:
```diff
@@ -0,0 +1,82 @@
+        self.n = num_words
+        self.concat_pos = concat_position
+        self.concat_function = self.get_concat_function()
+        self.spacy_model = spacy.load("en_core_web_sm")
```
`from initialize import spacy_nlp`
`self.spacy_model = spacy_nlp if spacy_nlp else spacy.load("en_core_web_sm")`

saqibns replied:
I'll make this change.

kaustubhdhole commented on transformations/word_noise/transformation.py:
```diff
@@ -0,0 +1,82 @@
+        )
+        noise = " ".join(noise_words)
+
+        context = self.concat_function(context, noise)
```
Maybe you could add a third concat function "both" ?

saqibns replied:
Yep. That sounds like a good idea!

kaustubhdhole commented:
Hi @saqibns this looks fine to me as it satisfies the review criteria and the code is well written. I've added a few comments.

On the same lines - it might be also be worth exploring appending sentences which make grammatical sense rather than just noise! Maybe replacing the answer entities & question verbs with different values and using them for replacing the answer sentence in the passage? I think the adversarial SQUAD paper do something like that. 

Thank you very much for your efforts!

saqibns commented:
Hi @kaustubhdhole thank you for your review and comments! I will update the PR according to your comments.

Yes, you are right about Adversarial SQUAD generating sentences that modify the question and append the generated sentence to the context. I checked their code and the script is available [here](https://github.com/robinjia/adversarial-squad/blob/master/src/py/convert_questions.py). The script is rather long, so should I make a new pull request for the same? 

Also, they mention in their paper that the custom rules they've defined to convert a question into a statement are not exhaustive and sometimes they yield sentences that aren't grammatically correct. Not that it should be a problem, but just mentioning it here.

kaustubhdhole commented:
Okay, sounds good - a new PR would be better as there is another PR too in making here https://github.com/GEM-benchmark/NL-Augmenter/pull/165

saqibns commented:
Hi @kaustubhdhole I have added the option of both appending and prepending noise to the context. I have also changed the code for `spacy` initialization as per your request.

kaustubhdhole commented on transformations/word_noise/transformation.py:
```diff
@@ -0,0 +1,101 @@
+from random import Random
```
I think the test cases in main are failing because of this. Can you use np.random and create a PR to fix that please?

saqibns replied:
Sure thing.

kaustubhdhole replied:
I created a temporary fix here: https://github.com/GEM-benchmark/NL-Augmenter/pull/195/files

saqibns replied:
Thank you!
